### ISSUE TITLE
Incorrectly encoded chars breaking response.data

### ISSUE THREAD

**Issue by finetuning-mg** [silkimen/cordova-plugin-advanced-http#160]
Hello,
I am having troubles with the response data of some POST calls (using the native client inside an Ionic 3 project).
I am setting up the POST call with: 
headers = { 'Content-Type': 'application/json; charset=UTF-8' };
http.setDataSerializer('utf8');
The response I get back is a JSON object, which sometimes include some weird characters like the following: {"field1": "aaaa", "field2": "bbbb", "field3":"Guasto gi� chiuso"}, where the � should be an 'à'.
In this case, the promise of the POST returns the response.data with the broken JSON string ' chiuso"}', causing the Javascript JSON parser to stop with an exception.
The people managing the server says the response is encoded as UTF-8, but I have the feeling it might be ISO-8559-1 instead.
Any idea on how to handle the situation? Could this be a bug of the native client?
This is currently happening when I run the project on Android (not yet checked on iOS).

Thanks

**Comment by finetuning-mg** [comment]
Update: the project works just fine when deployed to iOS (iPhone 8 with iOS 12.1). It seems to be an issue with the android part of the native plugin. Maybe the default encoding is something else than UTF-8?

**Comment by silkimen** [comment]
Hi finetuning-mg, when you are logging your `respose.data` you are receiving ` chiuso"}` instead of `{"field1": "aaaa", "field2": "bbbb", "field3":"Guasto già chiuso"}`? Did I understand your issue correctly?

**Comment by finetuning-mg** [comment]
Yes, please have a look at the debug from Google Chrome attached here. Still haven't been able to figure out if the string is incorrectly encoded by the server.
<img width="826" alt="screen shot 2018-11-23 at 13 04 21" src="https://user-images.githubusercontent.com/6928109/48986124-d92a5100-f110-11e8-8922-bd15b4bfdd4b.png">

**Comment by finetuning-mg** [comment]
Update on the topic. We have been able to solve the problem by adding the charset info in the response content-type header. Now it is something like:
`Content-Type: application/json;charset=utf-8`
and the native client is able to deal with data content just fine.
I suppose the native Java client is treating the response as a ISO-8859-1 string, and any UTF-8 encoded character is triggering some sort of exception or error. 
Maybe it could be useful to add a way to tell the native client which encoding is being used for the response?

**Comment by silkimen** [comment]
Ok thanks for this info, now I do remember why this problem is occurring!

Actually this plugin supports both UTF-8 and LATIN1 (which is ISO-8859-1). But if the charset is not given in the response, we're trying to guess it (see [CordovaHttp.java#L215-L217](https://github.com/silkimen/cordova-plugin-advanced-http/blob/master/src/android/com/synconset/cordovahttp/CordovaHttp.java#L215-L217)) by running some try-and-error code (see [CordovaHttp.java#L222-L236](https://github.com/silkimen/cordova-plugin-advanced-http/blob/master/src/android/com/synconset/cordovahttp/CordovaHttp.java#L222-L236)). This is working fine for most cases, but it's not 100% safe.

The best bet is to include the charset in the response. So you've done it absolutely correctly!

**Comment by pserna95** [comment]
I'm facing the same problem but with an extra issue which is that I'm making a GET request to an external API which I can't control so I'm not able to configure the charset info in the response content-type header.

I tried modifying CordovaHttp.java to make the plugin only accept UTF8 charset but I'm getting MalformedInputException - "Could not decode response data due to malformed data".

Any ideas?

**Comment by silkimen** [comment]
You can try to interpret the data yourself via [TextDecoder API](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder). You'll need to request the data as ArrayBuffer, see also #310.

**Comment by pserna95** [comment]
sendRequest method is not available with ionic-native/http@4.20.0. If I update to ionic-native/http@5.x.x I cannot import the module inside my providers array
